core/save-xml.c

```diff
@@ -798,3 +798,64 @@ int export_dives_xslt(const char *filename, const bool selected, const int units
 
 	return res;
 }
+
+void save_dive_sites_buffer(struct membuffer *b, const bool select_only, bool anonymize)
+{
+	int i;
+	put_format(b, "<divesites program='subsurface' version='%d'>\n", DATAFORMAT_VERSION);
+
+	/* save the dive sites - to make the output consistent let's sort the table, first */
+	sort_dive_site_table(&dive_site_table);
+	purge_empty_dive_sites(&dive_site_table);
+	for (i = 0; i < dive_site_table.nr; i++) {
+		struct dive_site *ds = get_dive_site(i, &dive_site_table);
+		/* Only write used dive sites when exporting selected dives */
+		if (select_only && !is_dive_site_used(ds, true))
+				continue;
+
+		put_format(b, "<site uuid='%8x'", ds->uuid);
+		show_utf8_blanked(b, ds->name, " name='", "'", 1, anonymize);
+		put_location(b, &ds->location, " gps='", "'");
+		show_utf8_blanked(b, ds->description, " description='", "'", 1, anonymize);
+		put_format(b, ">\n");
+		show_utf8_blanked(b, ds->notes, "  <notes>", " </notes>\n", 0, anonymize);
+		if (ds->taxonomy.nr) {
+			for (int j = 0; j < ds->taxonomy.nr; j++) {
+				struct taxonomy *t = &ds->taxonomy.category[j];
+				if (t->category != TC_NONE && t->value) {
+					put_format(b, "  <geo cat='%d'", t->category);
+					put_format(b, " origin='%d'", t->origin);
+					show_utf8_blanked(b, t->value, " value='", "'/>\n", 1, anonymize);
+				}
+			}
+		}
+		put_format(b, "</site>\n");
+	}
+	put_format(b, "</divesites>\n");
+}
+
+int save_dive_sites_logic(const char *filename, const bool select_only, bool anonymize)
+{
+	struct membuffer buf = { 0 };
+	FILE *f;
+	int error = 0;
+
+	save_dive_sites_buffer(&buf, select_only, anonymize);
+
+	if (same_string(filename, "-")) {
+		f = stdout;
+	} else {
+		try_to_backup(filename);
+		error = -1;
+		f = subsurface_fopen(filename, "w");
+	}
+	if (f) {
+		flush_buffer(&buf, f);
+		error = fclose(f);
+	}
+	if (error)
+		report_error("Save failed (%s)", strerror(errno));
```

Comment:
I note that we don't translate the error messages. Oh well, I guess that's not terrible...
Maybe a better error message `"Failed to save divesites to %s (%s)", filename, strerror(errno)`

Reply:
I updated the error messages and added the translate function.